library/core/src/array/mod.rs

```diff
@@ -564,15 +525,97 @@ impl<T, const N: usize> [T; N] {
+        // items.
+        unsafe { collect_into_array_unchecked(&mut self.iter_mut()) }
+    }
+}
+
+/// Pulls `N` items from `iter` and returns them as an array. If the iterator
+/// yields fewer than `N` items, this function exhibits undefined behavior.
+///
+/// If `iter.next()` panicks, all items already yielded by the iterator are
+/// dropped.
+///
+///
+/// # Safety
+///
+/// It is up to the caller to guarantee that `iter` yields at least `N` items.
+/// Violating this condition causes undefined behavior.
+unsafe fn collect_into_array_unchecked<I, const N: usize>(iter: &mut I) -> [I::Item; N]
+where
+    I: Iterator,
+{
+    match collect_into_array(iter) {
+        Some(array) => array,
+        // SAFETY: covered by the function contract.
+        None => unsafe { crate::hint::unreachable_unchecked() },
+    }
+}
+
+/// Pulls `N` items from `iter` and returns them as an array. If the iterator
+/// yields fewer than `N` items, `None` is returned and all already yielded
+/// items are dropped.
+///
+/// If `iter.next()` panicks, all items already yielded by the iterator are
+/// dropped.
+fn collect_into_array<I, const N: usize>(iter: &mut I) -> Option<[I::Item; N]>
+where
+    I: Iterator,
+{
+    if N == 0 {
+        // FIXME: is there a better way to achieve this?
+
+        // SAFETY: since N is 0, the array is empty and there are no
+        // uninitialized elements.
+        let out = unsafe { MaybeUninit::array_assume_init(MaybeUninit::uninit_array::<N>()) };
+        return Some(out);
+    }
+
+    struct Guard<T, const N: usize> {
+        ptr: *mut T,
+        initialized: usize,
+    }
+
+    impl<T, const N: usize> Drop for Guard<T, N> {
+        fn drop(&mut self) {
+            debug_assert!(self.initialized <= N);
+
+            let initialized_part = crate::ptr::slice_from_raw_parts_mut(self.ptr, self.initialized);
+
+            // SAFETY: this raw slice will contain only initialized objects.
+            unsafe {
+                crate::ptr::drop_in_place(initialized_part);
+            }
         }
+    }
 
-        // SAFETY: All elements of `dst` are properly initialized and
-        // `MaybeUninit<T>` has the same layout as `T`, so this cast is valid.
-        unsafe { (&mut out as *mut _ as *mut [&mut T; N]).read() }
+    let mut array = MaybeUninit::uninit_array::<N>();
+    let mut guard: Guard<_, N> =
+        Guard { ptr: MaybeUninit::slice_as_mut_ptr(&mut array), initialized: 0 };
+
+    while let Some(item) = iter.next() {
+        // SAFETY: `guard.initialized` starts at 0, is increased by one in the
+        // loop and the loop is aborted once it reaches N (which is
+        // `array.len()`). Therefore this index operation is safe.
+        unsafe {
+            array.get_unchecked_mut(guard.initialized).write(item);
+        }
+        guard.initialized += 1;
+
+        // Check if the whole array was initialized.
+        if guard.initialized == N {
```

Comment:
So if there are still elements in the iterator those will not be processed? Is that something we need to note here?

Reply:
Oh yes, that seems like something I should specifically mention it in the docs! Yes, only N items are pulled from the iterator and the iterator can be used afterwards. This is to make the function more flexible (think implementing `Iterator::chunks`).

Reply:
Because this can be called repeatedly, consider whether it might make sense to have a third version of this that returns `-> Result<[T; N], array::IntoIter<T, N>>` -- that would be nice for anything that wants to look at a whole iterator in chunks, even if it's not an exact multiple of N.

(That could be a different PR, though.)

Reply:
In https://github.com/rust-lang/rust/issues/81615, `(usize, [MaybeUninit<T>; N])` was suggested as a way to return incomplete arrays. But I would prefer having that added in a separate PR.